Implemented finding equation of Ellipse using slope as parameter and another method for calculation of circumference of ellipse - continued

avishrivastava11 opened the pull request:
Implemented finding equation of Ellipse using slope as parameter.

When the rotated ellipse is supported, this calculation can still be used but the `_slope` keyword will not be needed: the slope will be an attribute of the rotated ellipse.

Fixes #2815
Fixes #6952
Closes #15053 

This PR uses the approach to finding equation of ellipse using slope, length of semi minor axis and length of semi major axis as inputs given [here](https://math.stackexchange.com/questions/108270/what-is-the-equation-of-an-ellipse-that-is-not-aligned-with-the-axis/646971#646971)
This could be an added functionality to the equation finding method in class `Ellipse`.
Thanks to @smichr  for providing the approach.

Note : This is a continuation of #15053 
Please take a look at this PR and suggest changes. I will be glad to implement them.
Thanks.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* geometry
   * Implemented private method for finding equation of Ellipse using `_slope` as parameter
<!-- END RELEASE NOTES -->

avishrivastava11 committed:
added finding eqn of ellipse using slope, circumference method

avishrivastava11 commented:
@smichr Can you take a look?

sidhantnagpal commented on sympy/geometry/ellipse.py:
```diff
@@ -402,8 +439,30 @@ def encloses_point(self, p):
+        Let the center of the ellipse be at C=(xc,yc)
+
+        Let the major axis be the line that passes through C with a slope of s;
+        points on that line are given by the zeros of L(x,y) = (y−yc − s*(x−xc))**2.
```
The docstring contains unicode minus signs.

avishrivastava11 replied:
@sidhantnagpal  Yes. You're right. I missed that. Corrected now. Please take a look.

avishrivastava11 committed:
removed unicode '-' signs from docstring.

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -305,6 +305,43 @@ def center(self):
         """
         return self.args[0]
 
+    def ellipse_circumference(self):
```
as for Ellipse.normal_lines, let's add a `prec` keyword to circumference and just put this code there as a special case when `all(i.is_number for i in (self.vradius, self.hradius))`.

avishrivastava11 replied:
@smichr  Can you clarify what you are trying to say here again please? (I couldn't understand you comment completely)

smichr replied:
Don't make a new method. Add a parameter `prec` to `circumference` and have it handled the same way that it is handled in the `normal_lines` routine to calculate the numerical approximation of normal lines (but in this case you will use it to calculate an approximation of the circumference). Feel free to ask again if this isn't clear.

```python
def circumference(self, prec=None):
    if prec is None or not all(i.is_number for i in (self.vradius, self.hradius)):
        return normal symbolic circumference
    else:
        return calculated value to given prec
```

avishrivastava11 replied:
@smichr  I don't think that'll be a good idea (not making a new method) because of these reasons:
1. `circumference` is a `property`, while this method brings out the numerical value of the circumference

(I tried that and it resulted in the following error :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> p1 = Point(0, 0)
>>> e1 = Ellipse(p1, 3, 1)
>>> e.circumference(prec=20)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Mul' object is not callable
```)

2. `circumference`  produces an `ellipse object` whereas this method gives the numerical value.
If we still forcibly convert that ( `circumference` ) to a `non-property`, it'll be done by removing `@property` statement from the beginning. Then it would result in an `ellipse object` when the command is (I did that too and this is how we would get the results :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> p1 = Point(0, 0)
>>> e1 = Ellipse(p1, 3, 1)
>>> from sympy import Point, Ellipse, N
>>> N(e1.circumference())
13.3648932205553
```)

Now this could be acceptable but then quite some tests would have to be changed. (I have no issues doing either one)


So what do you think I should do?

smichr replied:
Hmm... `normal_lines` is a function so there is some precedent for Ellipse's related objects to be functions. 

The more helpful (SymPy-wide) thing to do would be to apply this method to the evaluation of the corresponding elliptic function when conditions are right (and I don't recall under what conditions this approximation succeeds). 

An alternative would be to make a private method (not property) named `_circumference` which can silently go away once the evaluation of ellipticals is fixed. 

Or name your method `circumference_n`...the "ellipse" part is redundant.

At this [summary](http://www.numericana.com/answer/ellipse.htm) page there is an exact representation in terms of the Legendre polynomial. Does that have better evaluation characteristics? Does the elliptical know how to rewrite itself into a Legendre polynomial?

smichr replied:
> has been implemented in `mpmath`

Something must be different or else the issue #6952 has been fixed. There was a *very large* difference in times taken to compute the result...but that was 5 years ago.  @avishrivastava11 , can you see if there is much difference in the times for these two today?

```python
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
```

smichr commented:
As for the timings, I get:
```
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 30.85376501083374)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 5.502314567565918)
```

avishrivastava11 commented:
@smichr 
>As for the timings, I get:

Yes. I am getting around the same values.
Do you think this time difference is significant? and if not, then we skip adding the `ellipse_circumference()` method?
or if it does matter then we actually replace the earlier method by the new one?
What do you think?

avishrivastava11 commented:
for the record I'm getting these values :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse, N
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.6152262687683105)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.3025321960449219)
```

As you can see, it takes almost half the time for `ellipse_circumference()` to do the operation

jksuom commented:
It looks like cache has some effect on the timings. The results are different if the tests are repeated.
```
>>> from sympy import Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.6817669868469238)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.3008091449737549)
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.06257200241088867)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.13084006309509277)
>>> 
```

avishrivastava11 commented:
@jksuom
Yes. In-fact, it seems to have a very interesting effect 
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.3472440242767334)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.21382713317871094)
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.036071062088012695)
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.06837582588195801)
```
the first time `circumference.n()` is run, it is slower than `ellipse_circumference()`. But after that, it becomes faster. (see the example I've posted above)

@smichr @jksuom  What should be done? Should `ellipse_circumference()` be removed ?

smichr commented:
> Should `ellipse_circumference()` be removed ?

It appears that the original issue either did not take into account the effect of cache or the method has been improved since then. I think the original issue should be closed and this routine removed. One final test would be to start and interactive session and run one of the timing tests. Then start a new session and run the other. If they don't differ by a factor of 2 or more, let's just close the issue.

avishrivastava11 commented:
@smichr
I tried both the methods in different sessions. Here are the findings : (I performed 3 trials)

Trial 1 :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 1.041156530380249)
>>> 
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.3112199306488037)
```

Trial 2 :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.61279296875)
>>> 
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.25516581535339355)
```

Trial 3 :
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).ellipse_circumference(), time()-t
(13.3648932205553, 0.3810257911682129)
>>> 
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.6 (default, Sep 12 2018, 18:26:19) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Point, Ellipse
>>> from time import time
>>> t = time();Ellipse((0,0),3,1).circumference.n(), time()-t
(13.3648932205553, 0.6159305572509766)
```

Well ! They're having a difference of nearly `>= 2`.
Should I remove the method `ellipse_circumference` still?
I think working on replacing the original `circumference` method would result in a very mediocre advantage of a small time difference (advantage).

Still, what would you recommend?
(I could remove the circumference part from this PR and we can continue with the rest of the PR)

smichr commented:
Unless you can trace down and find that mpmath is doing something different, I would remove the circimference method.

avishrivastava11 committed:
removed 'ellipse_circumference' method from ellipse.py

avishrivastava11 commented:
@smichr  I've removed the method `ellipse_circumference`. Please take a look. (I think the rest of the code is good to go)

smichr committed:
update docstring; remove unused import/parameter

The reference to the method used is given so there is no need for repeating the method in the docstring.
Demonstrate the rotation with an example.

smichr commented:
hopefully friendly and agreeable edits were made...

avishrivastava11 commented:
@smichr  Yes. The changes look great! Thanks for the inputs :)

smichr commented:
The last thought I have on this is to make the parameter `slope` private so it can be removed once the rotated ellipse is supported. Otherwise (when the rotated ellipse is implemented) this will have to be deprecated. What do you think?

avishrivastava11 committed:
making parameter 'slope' private

avishrivastava11 commented:
@smichr I wasn't exactly clear about "the process to make a parameter private". So I did some searching and made a commit. Please take a look.

smichr commented:
Looks good. I edited the OP and release notes.
